lib/bolt/transport/winrm/connection.rb

```diff
@@ -140,12 +142,59 @@ def mkdirs(dirs)
         end
 
         def write_remote_file(source, destination)
+          if target.options['file-protocol'] == 'smb'
+            write_remote_file_smb(source, destination)
+          else
+            write_remote_file_winrm(source, destination)
+          end
+        end
+
+        def write_remote_file_winrm(source, destination)
           fs = ::WinRM::FS::FileManager.new(@connection)
           fs.upload(source, destination)
         rescue StandardError => e
           raise Bolt::Node::FileError.new(e.message, 'WRITE_ERROR')
         end
 
+        def write_remote_file_smb(source, destination)
+          win_dest = destination.tr('/', '\\')
+          if (md = win_dest.match(/^([a-z]):\\(.*)/i))
+            # if drive, use admin share for that drive, so path is '\\host\C$'
+            path = "\\\\#{@target.host}\\#{md[1]}$"
+            dest = md[2]
+          elsif (md = win_dest.match(/^(\\\\[^\\]+\\[^\\]+)\\(.*)/))
+            # if unc, path is '\\host\share'
+            path = md[1]
+            dest = md[2]
+          else
+            raise ArgumentError, "Unknown destination '#{destination}'"
+          end
+
+          client = smb_client_login
```

Comment:
So even without caching / reusing the existing connection, this still provides significant benefit.

Did you happen to look at the impact of SMB connection reuse? (No worries if not, I believe there will be an effort to support better connection handling for winrm in the future, so this could come along for the ride)

Reply:
Minor terminology nit, the SMB connection is reused, but the tree is not. I believe the tree connect/disconnect is similar to concept to mounting a remote filesystem, and in this PR we do that for each file upload.

It'd be possible to reuse the same tree across multiple uploads, but bolt would need to keep track of which destination paths map to which trees, and create them as needed. It might be useful if there large numbers of small files being uploaded in a plan, but I didn't test it. Does bolt support uploading a directory?

Reply:
Thanks for the clarification - I completely missed the `@client` bit